src/join/core.jl

```diff
@@ -428,84 +509,306 @@ function _innerjoin_postprocess_int(left::AbstractVector{<:Union{Integer, Missin
+    sizehint!(dict, 2 * min(right_len, typemax(Int) >> 2))
+
+    right isa OnCol && _prehash(right)
+    left isa OnCol && _prehash(left)
+
+    for (idx_r, val_r) in enumerate(right)
+        haskey(dict, val_r) && return _semijoin_dup(left, right, dict, idx_r,
+                                                    seen_rows, Val(right_shorter))
+        dict[val_r] = idx_r
+    end
+
+    @inbounds for (idx_l, val_l) in enumerate(left)
+        # we know that dict contains only positive values
+        idx_r = get(dict, val_l, -1)
+        if idx_r != -1
+            if right_shorter
+                seen_rows[idx_l] = true
+            else
+                seen_rows[idx_r] = true
             end
         end
     end
+    return seen_rows
+end
 
-    disallow_sorted = false
+function _semijoin_unsorted_int(left::AbstractVector{<:Union{Integer, Missing}},
+                                right::AbstractVector{<:Union{Integer, Missing}},
+                                seen_rows::AbstractVector{Bool},
+                                ::Val{right_shorter}) where {right_shorter}
+    minv, maxv = extrema_missing(right)
 
-    for (lc, rc) in zip(left_cols, right_cols)
-        @assert length(lc) == left_len
-        @assert length(rc) == right_len
-        lc_et = nonmissingtype(eltype(lc))
-        rc_et = nonmissingtype(eltype(rc))
+    val_range = big(maxv) - big(minv)
+    if val_range > typemax(Int) - 3 || val_range ÷ 2 > max(64, length(right)) ||
+       minv < typemin(Int) + 2 || maxv > typemax(Int) - 3
+       return _semijoin_unsorted(left, right, seen_rows, Val(right_shorter))
+    end
 
-        # special case common safe scenarios when eltype between left and right column
-        # can be different or non-concrete
-        lc_et <: Real && rc_et <: Real && continue
-        lc_et <: AbstractString && rc_et <: AbstractString && continue
+    offset = 1 - Int(minv) # we are now sure it does not overflow
+    len = Int(maxv) - Int(minv) + 2
+    dict = zeros(Int, len)
 
-        # otherwise we require non-missing eltype of both sides to be the same and concrete
-        lc_et === rc_et && isconcretetype(lc_et) && continue
+    @inbounds for (idx_r, val_r) in enumerate(right)
+        i = ismissing(val_r) ? length(dict) : Int(val_r) + offset
+        if dict[i] > 0
+            return _semijoin_dup_int(left, right, dict, idx_r, offset, Int(minv), Int(maxv),
+                                     seen_rows, Val(right_shorter))
+        end
+        dict[i] = idx_r
+    end
 
-        # we disallow using sorted branch for some columns that theoretically
-        # could be safely sorted (e.g. having Any eltype but holding strings)
-        # for safety reasons assuming that such cases will be rare in practice
-        disallow_sorted = true
+    @inbounds for (idx_l, val_l) in enumerate(left)
+        if ismissing(val_l)
+            idx_r = dict[end]
+            if idx_r > 0
+                if right_shorter
```

Comment:
Did you make `right_shorter` a `Val` just because of this? I think LLVM should be able to hoist the check so using `Val` shouldn't be necessary.

Also, maybe something like `r  = idx_r > 0; seen_rows[idx_l] |= r` could be slightly faster if that avoids a branch? `val_l === missing` might also be faster than `ismissing(val_l)` (until Julia 1.7).

Reply:
Thank you for the comment:
* indeed `=== missing` is faster; I have changed this
* using `|= r` can avoid branch in only one of the cases, in the other we use `idx_r` also for indexing so it had to stay; I have changed this
* regarding `Val` - I will refactor the code so that we do not need it